moveit_core/cost_functions/src/cost_functions.cpp

```diff
@@ -0,0 +1,74 @@
+/*********************************************************************
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2023, PickNik Inc.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of PickNik Inc. nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Sebastian Jahr */
+
+#include <moveit/cost_functions/cost_functions.hpp>
+
+namespace moveit
+{
+namespace cost_functions
+{
+[[nodiscard]] ::planning_interface::StateCostFn
+getClearanceCostFn(moveit::core::RobotState& robot_state, const std::string& group_name,
+                   const planning_scene::PlanningSceneConstPtr& planning_scene)
+{
+  // Create cost function
+  return [robot_state, group_name, planning_scene](const Eigen::VectorXd& state_vector) mutable {
```

Comment:
There are multiple sources that discourage using std::bind (like [abseil](https://abseil.io/tips/108) and [turner](https://www.youtube.com/watch?v=ZlHi8txU4aQ)) because of the compile and runtime overhead it causes. Maybe C++20's  std::bind_front fixes that but since we're not using it at the moment I'd prefer the lambdas for now. I guess if I'd like to avoid the dependency on the planning_interface I could also just abstain from using the typedef and use `std::function<double(const Eigen::VectorXd& state_vector)>` instead of `StateCostFn`. So the dependency is not mandatory.

Reply:
What I mean with avoiding the dependency is that if you just have the function that computes the cost here, you can turn that into a StateCostFn or any other function signature at the caller side (e.g. stomp or other libraries may have their own cost function signature, but they could still use these cost computation functions and transform them into what they need via bind or lambdas. Of course this doesn't matter if we think that `double(Eigen::VectorXd)` is a general enough cost-definition function, which it probably is. I'm okay with leaving this as is for now and iterate later if needed, but see my comment below for the lambda captures.

On a separate note, I wish we had the absl dependency to be able to use absl::bind_front, StatusOr and others, and to follow their TotW and style guide.